subprojects/model-core/src/main/java/org/gradle/internal/extensibility/ConventionAwareHelper.java

```diff
@@ -60,14 +64,27 @@ private MappedProperty map(String propertyName, MappedPropertyImpl mapping) {
                 "You can't map a property that does not exist: propertyName=" + propertyName);
         }
 
-        if (_ineligiblePropertyNames.contains(propertyName)) {
+        // When there's a Property, use its `.convention()` method instead
+        // This is something we added to support properties migrated in the future from
+        // Java bean to Property where old code uses ConventionMapping to set conventions.
+        Class<? extends IConventionAware> sourceType = _source.getClass();
+        Method getter = JavaPropertyReflectionUtil.findGetterMethod(sourceType, propertyName);
```

Comment:
This is potentially expensive and I think it only applies when the propertyName is an "ineligible" property. You might want to instead put this into the `else if` condition below so we only check the return type when distinguishing between Property and Provider.

Alternatively, we could change `ineligible` to only include `Provider` types and create a new list that keeps track of `Property` types.  The list of `ineligible` properties is dynamically built through the ASM class generator to avoid this reflection.

Reply:
Follow up: https://github.com/gradle/gradle/pull/25194